### PR TITLE
Output guardrails per-guardrail evaluation

### PR DESCRIPTION
**Pull request by chaecramb**
Continuation of work on Trello: https://trello.com/c/0ssf3xuh/2376-port-output-guardrail-evaluations-to-the-python-evaluation-framework

#15 added evaluation for `output_guardrails` based on whether *any* guardrail was triggered (overall `triggered` status: True/False). This provided a high-level view but didn't allow for diagnosing the performance of individual guardrails. This PR enhances this by calculating and reporting metrics (Precision, Recall, F1-score) for each of the individual output guardrails, providing a more granular approach to understand which specific guardrails are performing well and which need improvement.

## Key changes include:

1.  **Updated Data Models:**
    *   `EvaluationResult` now directly consumes `expected_guardrails: dict[str, bool]` and `actual_guardrails: dict[str, bool]`.
    *   GOV.UK Chat output guardrails have been updated to provide the individual guardrail names and trigger status (https://github.com/alphagov/govuk-chat/pull/168), so that this data can be provide via the output guardrail evaluation rake task.
    *   The input model, `GenerateInput`, has also been updated to use `expected_guardrails: dict[str, bool]`.
2.  **Dynamic Guardrail Handling:**
    *   Guardrail names and their total count are now determined dynamically at runtime by inspecting the keys present in all `expected_guardrails` and `actual_guardrails` dictionaries across the evaluation dataset.
    *   Hardcoded assumptions about the number of guardrails and generic "G1...G7" labels are no longer used.
3.  **Metric Calculation:**
    *   `AggregateResults` compiles a global, sorted list of all unique `guardrail_names` encountered in the dataset.
    *   For each `EvaluationResult`, the `expected_guardrails` and `actual_guardrails` dictionaries are converted into binary vectors. These vectors are ordered according to the global `guardrail_names` list, ensuring consistency for metric calculations.
    *   `sklearn.metrics.precision_score`, `recall_score`, and `f1_score` are then used to compute the precision, recall, and F1 score for each individual guardrail across all evaluation results.
4.  **Output Update:**
    *   `AggregateResults.to_dict()` and `AggregateResults.for_csv()` now include these per-guardrail metrics in the evaluation output.
    *   Column headers in reports and console output use the actual discovered guardrail names (e.g., `Precision [appropriate_language]`, `Recall [sensitive_financial_matters]`, etc.), providing clearer and more direct insights.
5.  **Warning Removal:** Warning logic related to string parsing of guardrail outputs has been removed, as the new system relies on pre-structured data.

## Example output:

```
$ uv run govuk_chat_evaluation output_guardrails --guardrail_type=answer_guardrails --input_path="data/output_guardrails_no_generate.jsonl" --provider=openai --no-generate
Wrote results to results/output_guardrails/2025-05-07T09:16:18/results.csv
Wrote aggregates to results/output_guardrails/2025-05-07T09:16:18/aggregate.csv
Aggregate Results
---------------------------------------  --------
Evaluated                                6
Any-triggered Precision                  0.666667
Any-triggered Recall                     0.666667
Any-triggered True positives             2
Any-triggered True negatives             2
Any-triggered False positives            1
Any-triggered False negatives            1
Precision [appropriate_language]         0.4
Recall [appropriate_language]            0.666667
F1 [appropriate_language]                0.5
Precision [contains_pii]                 1
Recall [contains_pii]                    0.666667
F1 [contains_pii]                        0.8
Precision [illegal]                      1
Recall [illegal]                         0.666667
F1 [illegal]                             0.8
Precision [inappropriate_style]          1
Recall [inappropriate_style]             0.666667
F1 [inappropriate_style]                 0.8
Precision [political]                    1
Recall [political]                       0.666667
F1 [political]                           0.8
Precision [sensitive_financial_matters]  1
Recall [sensitive_financial_matters]     0.666667
F1 [sensitive_financial_matters]         0.8
Precision [unsupported_statements]       1
Recall [unsupported_statements]          0.666667
F1 [unsupported_statements]              0.8
---------------------------------------  --------

Wrote used config to results/output_guardrails/2025-05-07T09:16:18/config.yaml
```

### CODE REVIEW

File: govuk_chat_evaluation/output_guardrails/evaluate.py [hunk 1]
```diff
@@ -28,15 +34,125 @@ def classification(self) -> str:
...
+            guardrail_numbers = [int(num) for num in guardrails_str.split(",")]
+            # Check guardrail numbers are in range and unique
+            if not all(1 <= num <= num_guardrails for num in guardrail_numbers):
+                raise ValueError(
+                    f"Incorrect format for guardrail result: {exact_string}"
+                )
+            if len(guardrail_numbers) != len(set(guardrail_numbers)):
+                raise ValueError(
+                    f"Incorrect format for guardrail result: {exact_string}"
+                )
```
Comment (exfalsoquodlibet) [inline govuk_chat_evaluation/output_guardrails/evaluate.py:134]: Question: are we too strict here?

**First case - some out-of-range predicted indexes.**
If the LLM guardrail returns "True | "1,3,9", we currently discard the entire result because the "9" is outside the valid range. Instead of discarding the entire result, we could consider raising a warning for the invalid guardrail (i.e., "9"), but still proceed with the valid indices (i.e., "1" and "3").

**Second case - duplicate indeces.**
If the LLM guardrail returns "True | "1,3,3", we are currently discarding the entire result due to the duplicate "3". An alternative approach could be to raise a warning for the duplicate value and then proceed by de-duplicating the list of indices before continuing with the evaluation.

Comment (chaecramb) [reply govuk_chat_evaluation/output_guardrails/evaluate.py:134]: Yes, good point. I should distinguish between things that should give a warning, but can still be evaluated, and things that need to raise an error.

File: govuk_chat_evaluation/output_guardrails/evaluate.py [hunk 2]
```diff
@@ -67,8 +210,44 @@ def recall(self) -> float:
...
+        return f1_score(
+            expected_vectors,
+            actual_vectors,
+            average=None,  # type: ignore
+            zero_division=0,  # type: ignore
+        ).tolist()  # type: ignore
+
     def to_dict(self) -> dict[str, Any]:
-        return {
+        base_metrics = {
```
[the remark this answers is not in the document]
Comment (chaecramb) [reply govuk_chat_evaluation/output_guardrails/evaluate.py:250]: Yes, good point. I'll make that clearer.

File: govuk_chat_evaluation/output_guardrails/evaluate.py [hunk 3]
```diff
@@ -28,15 +34,125 @@ def classification(self) -> str:
...
+            raise ValueError(f"Incorrect format for guardrail result: {exact_string}")
+
+        triggered_str = match.group(1)
+        guardrails_str = match.group(2)  # comma-separated guardrailnumbers, if present
+        none_str = match.group(3)  # The literal string "None", if present
+
+        if triggered_str == "True":
+            if guardrails_str is None:
+                raise ValueError(
+                    f"Incorrect format for guardrail result: {exact_string}"
```
[the remark this answers is not in the document]
Comment (exfalsoquodlibet) [reply govuk_chat_evaluation/output_guardrails/evaluate.py:116]: Yes, that would be useful, so that we can review these cases and see what happened (e.g., why the Output Guardrail returned "True | None" and similar ).

File: govuk_chat_evaluation/logging_utils.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import logging
+from pathlib import Path
+
+
+def configure_run_logging(
```
Comment (kevindew) [inline govuk_chat_evaluation/logging_utils.py:5]: I think this opens up a bit of a can of worms that we might want to have a chat about how we can deal with.

The can of worms is that we're already doing a bunch of logging by writing to stdout and we should be consistent with logging as to whether it's written to a logger, stdout or a log file.

Off the top of my head I'd wonder if we'd want to extract this out of this PR and then apply a logger to all the techniques at the same time. But lets have a chat about it to work out what is most effective.

(Awesome work on this PR btw, I've still got much to review but thought I'd send this comment first since it could be a bit significant and I've got calendar commitments about to start)

Comment (chaecramb) [reply govuk_chat_evaluation/logging_utils.py:5]: Good point, that makes sense.

File: govuk_chat_evaluation/output_guardrails/evaluate.py [hunk 5]
```diff
@@ -28,15 +42,184 @@ def classification(self) -> str:
...
+        else:  # triggered_str == "False"
+            if none_str != "None" or guardrails_str is not None:
+                raise ValueError(
+                    f"Guardrail string '{exact_string}' starts with 'False' but is not "
+                    f"followed by 'None'. Expected format 'False | None'."
+                )
+            # Return non-triggered result with empty numbers/warnings
+            return _GuardrailParseResult(
+                is_triggered=False, valid_numbers=[], warnings=[]
+            )
```
Comment (kevindew) [inline govuk_chat_evaluation/output_guardrails/evaluate.py:211]: It does feel like we've got a lot of logic for error handling. I don't think we need all of it because a bunch would be handled in the Rails app: https://github.com/alphagov/govuk-chat/blob/beba043922091350b0ef38bf23650dd453e53e78/lib/guardrails/multiple_checker.rb#L89-L93

We could take the approach that whatever input we have we create a GuardrailParseResult object so then we don't have to deal with it being None.

Then I think warning wise we could simplify it to:

- a single warning if regex is not matched (in which case we have triggered false and valid as an empty list), we could also make regex stricter to match `True \| (numbers spaces and commas) | False \| None`
- once we've got the numbers parsed (`numbers = [int(x.strip()) for x in s.split(',')]`):
  - any unexpected numbers
  - absence of valid numbers
  - any duplicates

I think that would be all we'd need to check?

File: govuk_chat_evaluation/output_guardrails/evaluate.py [hunk 6]
```diff
@@ -67,17 +277,66 @@ def recall(self) -> float:
...
+    def precision_per_guardrail(self) -> List[float]:
+        """Calculates precision for each guardrail individually."""
+        expected_vectors, actual_vectors = self._expected_actual_vectors
+        if not expected_vectors:  # Avoid calling score with empty lists
+            return [np.nan] * NUM_GUARDRAILS
+        return precision_score(
+            expected_vectors,
+            actual_vectors,
+            average=None,  # type: ignore
+            zero_division=0,  # type: ignore
```
Comment (kevindew) [inline govuk_chat_evaluation/output_guardrails/evaluate.py:289]: we have a convention in the rest of this repo to use nan rather than zero, could we use that here?

Edit: same for recall and f1_score

File: govuk_chat_evaluation/output_guardrails/evaluate.py [hunk 7]
```diff
@@ -47,14 +230,41 @@ def __init__(self, evaluation_results: List[EvaluationResult]):
...
+            expected_vectors.append(
+                expected_vec if expected_vec is not None else zero_vector
+            )
+            actual_vectors.append(actual_vec if actual_vec is not None else zero_vector)
+
+        # Handle the edge case where there are no results to evaluate
+        if not expected_vectors:
+            return [], []
+
+        return expected_vectors, actual_vectors
```
Comment (kevindew) [inline govuk_chat_evaluation/output_guardrails/evaluate.py:266]: This method feels quite complex. I imagine if we always have a guardrail_parse result this could be simpler.

File: govuk_chat_evaluation/output_guardrails/evaluate.py [hunk 8]
```diff
@@ -47,14 +230,41 @@ def __init__(self, evaluation_results: List[EvaluationResult]):
...
+            actual_vec = result.actual_exact_triggered
+
+            expected_vectors.append(
+                expected_vec if expected_vec is not None else zero_vector
+            )
+            actual_vectors.append(actual_vec if actual_vec is not None else zero_vector)
+
+        # Handle the edge case where there are no results to evaluate
+        if not expected_vectors:
+            return [], []
```
Comment (kevindew) [inline govuk_chat_evaluation/output_guardrails/evaluate.py:264]: I'm not sure why we need this as expected_vectors and actual_vectors would already be empty lists?

**Review by kevindew — COMMENTED** [review]
Nice job Chae, this all looked like some rather serious coding!

I've made some rather bold suggestions, I hope too frustrating, I think there's a bunch we can simplify in this that should make it a bit easier to comprehend and to test. 

Happy to have a chat about it. Broadly I think we should cut out the logger and just store the warnings now and have a follow up PR about adding the logger and then move parsing into GuardrailParseResult and handle fewer errors.

**Comment by chaecramb** [comment]
@kevindew Since it was quite a big refactor, I've addressed your comments with a new commit rather than rebasing, to make the changes clearer.

**Comment by chaecramb** [comment]
@kevindew This has been through another major refactor. It should be much simpler now, as all parsing logic, and the error handling that went along with it has been stripped out, now relying on the Chat codebase for that.

File: govuk_chat_evaluation/output_guardrails/evaluate.py [hunk 9]
```diff
@@ -30,55 +32,113 @@ def classification(self) -> str:
...
+                (
+                    int(evaluation_result.expected_triggered),
+                    int(evaluation_result.actual_triggered),
+                )
+                for evaluation_result in self.evaluation_results
             )
-            for evaluation_result in self.evaluation_results
-        ]
-        expected, actual = zip(*pairs_list)
+        )
```
Comment (kevindew) [inline govuk_chat_evaluation/output_guardrails/evaluate.py:67]: I'm not too sure why we changed this? it's probably a little easier to understand in it's previous form.

File: govuk_chat_evaluation/output_guardrails/evaluate.py [hunk 10]
```diff
@@ -30,55 +32,113 @@ def classification(self) -> str:
...
+        guardrail_set = {
+            name
+            for result in evaluation_results
+            for name in list(result.expected_guardrails.keys())
+            + list(result.actual_guardrails.keys())
+        }
+        self.guardrail_names: List[str] = sorted(guardrail_set)
+
     @cached_property
     def _expected_actual_lists(self) -> tuple[List[int], List[int]]:
```
Comment (kevindew) [inline govuk_chat_evaluation/output_guardrails/evaluate.py:58]: should we rename this to expected_actual_triggered_lists to associate it with triggered?

File: govuk_chat_evaluation/output_guardrails/evaluate.py [hunk 11]
```diff
@@ -30,55 +32,113 @@ def classification(self) -> str:
...
+
+        expected_vectors = [
+            to_vector(result.expected_guardrails) for result in self.evaluation_results
+        ]
+        actual_vectors = [
+            to_vector(result.actual_guardrails) for result in self.evaluation_results
+        ]
+        return expected_vectors, actual_vectors
+
+    def _metric_any(self, metric_function):
```
Comment (kevindew) [inline govuk_chat_evaluation/output_guardrails/evaluate.py:87]: shall we name this _triggered_metric to reflect it's usage for triggered?

File: govuk_chat_evaluation/output_guardrails/evaluate.py [hunk 12]
```diff
@@ -30,55 +32,113 @@ def classification(self) -> str:
...
+                for evaluation_result in self.evaluation_results
             )
-            for evaluation_result in self.evaluation_results
-        ]
-        expected, actual = zip(*pairs_list)
+        )
         return list(expected), list(actual)
 
+    @cached_property
+    def _expected_actual_vectors(self) -> tuple[List[List[int]], List[List[int]]]:
```
Comment (kevindew) [inline govuk_chat_evaluation/output_guardrails/evaluate.py:71]: shall we include guardrails in the name of this?

File: govuk_chat_evaluation/output_guardrails/evaluate.py [hunk 13]
```diff
@@ -30,55 +32,113 @@ def classification(self) -> str:
...
+        ]
+        actual_vectors = [
+            to_vector(result.actual_guardrails) for result in self.evaluation_results
+        ]
+        return expected_vectors, actual_vectors
+
+    def _metric_any(self, metric_function):
+        expected, actual = self._expected_actual_lists
+        if not expected:
+            return np.nan
```
Comment (kevindew) [inline govuk_chat_evaluation/output_guardrails/evaluate.py:90]: I've got a feeling I've asked this before so apologies if you have deja vu, but I'm not sure why we need this condition as I'd have thought the metric function itself would generate nan on an empty input scenario?

File: govuk_chat_evaluation/output_guardrails/evaluate.py (hunk 8 above)
Comment (kevindew) [reply govuk_chat_evaluation/output_guardrails/evaluate.py:74]: Sorry if we cleared this up in a chat, but I don't quite remember why we need this conditional and would have thought it behaves the same without this?

File: govuk_chat_evaluation/output_guardrails/evaluate.py [hunk 14]
```diff
@@ -30,55 +32,113 @@ def classification(self) -> str:
...
+        return metric_function(expected, actual, zero_division=np.nan)  # type: ignore
+
+    def _metric_per_guardrail(self, metric_function):
+        expected_vectors, actual_vectors = self._expected_actual_vectors
+        if not expected_vectors:
+            return [np.nan] * len(self.guardrail_names)
```
Comment (kevindew) [inline govuk_chat_evaluation/output_guardrails/evaluate.py:96]: I'm not too clear on this conditional too - as we're only checking expected. I somewhat wonder is it even possible to get an empty expected vector? is that the case when there are zero guardrails?

File: tests/output_guardrails/test_evaluate.py [hunk 15]
```diff
@@ -5,6 +5,7 @@
 import numpy as np
 import pytest
 import logging
+from pytest import approx
```
Comment (kevindew) [inline tests/output_guardrails/test_evaluate.py:8]: this looks very handy

**Review by kevindew — COMMENTED** [review]
Nice work Chae. This looks nicely simpler.

I think there's just a few things to focus on:

- consistent naming in the AggregateResults class
- working out some of the edge case conditionals
- using lowercase dict and list for types in evaluate.py

File: govuk_chat_evaluation/output_guardrails/evaluate.py [hunk 16]
```diff
@@ -14,12 +14,14 @@
 
 class EvaluationResult(BaseModel):
     question: str
-    expected_outcome: bool
-    actual_outcome: bool
+    expected_triggered: bool
+    actual_triggered: bool
+    expected_guardrails: Dict[str, bool]
+    actual_guardrails: Dict[str, bool]
```
[the remark this answers is not in the document]
Comment (chaecramb) [reply govuk_chat_evaluation/output_guardrails/evaluate.py:20]: Ah I see, apparently this is a legacy thing I picked up, required for compatibility with Python version before 3.9.

File: govuk_chat_evaluation/output_guardrails/evaluate.py (hunk 13 above)
Comment (chaecramb) [reply govuk_chat_evaluation/output_guardrails/evaluate.py:90]: I put this in after creating an accidental edge case in my tests where I had an empty expected vector, but content in the actual vector - it breaks in that scenario rather than returning nan. However that was probably a bit artificial. I only created the scenario by incorrectly setting up my tests and it shouldn't happen in reality unless someone has messed up their input data, in which case we probably want it to error anyway, so I'll just remove this guard.

File: govuk_chat_evaluation/output_guardrails/evaluate.py (hunk 14 above)
Comment (chaecramb) [reply govuk_chat_evaluation/output_guardrails/evaluate.py:96]: This was following the pattern I established above, but I think you're correct to call it out, as an empty expected vector shouldn't be an issue in reality.

File: govuk_chat_evaluation/output_guardrails/evaluate.py (hunk 8 above)
Comment (kevindew) [reply govuk_chat_evaluation/output_guardrails/evaluate.py:75]: I think i forgot to unresolve this comment so you may have missed it.

**Review by kevindew — APPROVED** [review]
Great - just a small thing but I don't think it needs another review.
